### ISSUE TITLE
groot/{cmd/root-gen-rfunc,rtree/rfunc}: additional rfuncs, rfunc name cosmetics

### ISSUE THREAD

**Issue by sbinet** [go-hep/hep#737]
raised in https://github.com/go-hep/hep/pull/734#issuecomment-634926739 by @rmadar:

---
Thanks a lot for this, this is really *super* useful! I would like to propose few signatures/types that I personally find useful (feel free to tell me if this is too specific). But before:
 1. I would like to know: is it somehow "expensive" to generate many pre-defined functions or not?
 2. It would be nice to have the list of implemented type/signatures in the doc. Right now, I see something like this:
![image](https://user-images.githubusercontent.com/30081646/83069412-21aa2e00-a06a-11ea-8c47-43f455142749.png)
But I think we might want to tune the naming convention to make appear input type/multiplicity and output type. Why not something like: `In<T1>Ar<N1><T2>Ar<N2>...Out<T1>Ar<N1><T2>Ar<N2>`. For `func(in32, float64, float64) bool`, it would look like `InI32Ar1F64Ar1F64Ar1OutB` ... ok maybe not so readable (it would be better with `_` in this case ;-) )!

### Wish-list and motivation

My strategy is to work only with `float64` (for values) and `bool` (for cuts) as output types, to avoid too much combination. Enable pre-defined functions for slices would be also nice.  Here is the (first) full wish-list, organized by goals:

**1. Other types to `float64` conversion**
  + `func(float32) float64`
  + `func(int) float64`
  + `func(int32) float64`
  + `func(int64) float64`

**2. "Multi-branches" cuts** (I guess it's fair to let the "mixed" signatures to the user)
  + `func(int) bool` + arrities
  + `func(int32) bool` + arrities
  + `func(float32) bool` + arrities
  + `func(float64) bool` + arrities

**3. Slice manipulations** (both for reducing operation and element operation)
  + `func([]<input>) []float64`
  + `func([]<input>) float64`
  + `func([]<input>, int) float64`
(where `<input>` can be `int`, `int32`, `int64`, `float32` or `float64`)
---

- [x] change default naming convention
- [x] add more rfunc formula implementations

**Comment by sbinet** [comment]
> I would like to know: is it somehow "expensive" to generate many pre-defined functions or not?

with the new scheme (_ie:_ sans the automagical registration of pre-defined formulae), it's less expansive.
_e.g.:_
```
$> ll go-hep.org/x/hep/groot/rtree/rfunc.a 
-rw-r--r-- 1 binet binet 1.4M May 28 08:50 go-hep.org/x/hep/groot/rtree/rfunc.a

$> ll go-hep.org/x/hep/groot/rtree.a 
-rw-r--r-- 1 binet binet 3.7M May 28 08:50 go-hep.org/x/hep/groot/rtree.a
```
of course, some balance and restrain have to be applied: we don't want to bloat too much that little `rfunc` package :)

> But I think we might want to tune the naming convention to make appear input type/multiplicity and output type

yeah. the naming/mangling scheme could be improved, indeed.
what about:
- `func(float64)float64` -> `FuncDtoD`
- `func(x1,x2 float64) float64` -> `FuncDDtoD`
- `func(int32,float32,bool) float64` -> `FuncIFBtoD`
(well, the `root-gen-rfunc` tool allows users to specify the output name as they wish)

> 1. Other types to float64 conversion
> `func(float32) float64`
> `func(int) float64`
> `func(int32) float64`
> `func(int64) float64`

I am ok with adding those to `rfunc`, except for `func(int)float64`.
`int` shouldn't be stored in a `rtree.Tree`, so I don't see much value in having that one generated. (branches/leaves are always sized types)

> 2. "Multi-branches" cuts

same comment _wrt_ `int`.
how much arity would you want: 2,3,4,5 ? 10? let's start with 5?

> 3. Slice manipulations

same comment _wrt_ `int` as `<input>`.
also, for `func([]<input>, int) float64`, where there's the same argument.

**Comment by rmadar** [comment]
Your naming convention is quite nice! One can add 's' for slices I guess. 

Ok for int, I didn't know

An arity of 5 should be more than enough! 

Thanks!

**Comment by sbinet** [comment]
Do you want to give it a stab?
(At least the list of functions you'd like to add. The naming convention also if you want)

**Comment by rmadar** [comment]
Sure, I'll give a shot!

**Comment by sbinet** [comment]
I'll send a PR tackling the (default) naming convention change.

**Comment by sbinet** [comment]
done. I had to change a bit what I initially proposed because of possible collisions between slices and uint16.
see:
- https://godoc.org/go-hep.org/x/hep/groot/rtree/rfunc#FuncF32F32F32ToF32
- https://godoc.org/go-hep.org/x/hep/groot/rtree/rfunc#FuncF32sToF64s

**Comment by sbinet** [comment]
fixed w/ #742.
